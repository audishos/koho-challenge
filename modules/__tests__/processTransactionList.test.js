const path = require('path');
const processTransactionList = require('../processTransactionList');
const { DAILY_AMOUNT_LIMIT } = require('../constants');

const testName = path.parse(__filename).name;

const mockTransactionList = [
  // accepted
  {
    id: '1',
    customer_id: '1',
    load_amount: DAILY_AMOUNT_LIMIT / 4,
    time: Date.parse('2020-04-01'),
  },
  // accepted
  {
    id: '2',
    customer_id: '1',
    load_amount: DAILY_AMOUNT_LIMIT / 4,
    time: Date.parse('2020-04-01'),
  },
  // accepted
  {
    id: '3',
    customer_id: '1',
    load_amount: DAILY_AMOUNT_LIMIT / 4,
    time: Date.parse('2020-04-01'),
  },
  // rejected - daily transaction count limit exceeded
  {
    id: '4',
    customer_id: '1',
    load_amount: DAILY_AMOUNT_LIMIT / 4,
    time: Date.parse('2020-04-01'),
  },
  // rejected - load amount exceeds daily limit
  {
    id: '420',
    customer_id: '8',
    load_amount: DAILY_AMOUNT_LIMIT + 1,
    time: Date.parse('2020-04-01'),
  },
  // accepted
  {
    id: '666',
    customer_id: '40',
    load_amount: DAILY_AMOUNT_LIMIT / 2 - 1,
    time: Date.parse('2020-04-01'),
  },
  // accepted
  {
    id: '675',
    customer_id: '40',
    load_amount: DAILY_AMOUNT_LIMIT / 2 - 1,
    time: Date.parse('2020-04-01'),
  },
  // rejected - duplicate transaction
  {
    id: '675',
    customer_id: '40',
    load_amount: DAILY_AMOUNT_LIMIT / 2 - 1,
    time: Date.parse('2020-04-01'),
  },
  // rejected - load amount exceeds daily limit
  {
    id: '686',
    customer_id: '40',
    load_amount: 3,
    time: Date.parse('2020-04-01'),
  },
  // accepted
  {
    id: '728',
    customer_id: '123',
    load_amount: DAILY_AMOUNT_LIMIT,
    time: Date.parse('2020-04-01'),
  },
  // accepted
  {
    id: '734',
    customer_id: '123',
    load_amount: DAILY_AMOUNT_LIMIT,
    time: Date.parse('2020-04-02'),
  },
  // accepted
  {
    id: '756',
    customer_id: '123',
    load_amount: DAILY_AMOUNT_LIMIT,
    time: Date.parse('2020-04-03'),
  },
  // accepted
  {
    id: '757',
    customer_id: '123',
    load_amount: DAILY_AMOUNT_LIMIT,
    time: Date.parse('2020-04-04'),
  },
  // rejected - load amount exceeds weekly limit
  {
    id: '758',
    customer_id: '123',
    load_amount: DAILY_AMOUNT_LIMIT,
    time: Date.parse('2020-04-05'),
  },
  // accepted
  {
    id: '1025',
    customer_id: '5',
    load_amount: DAILY_AMOUNT_LIMIT - 1,
    time: Date.parse('2020-04-06'),
  },
];

describe(testName, () => {
  const results = processTransactionList(mockTransactionList);

  it('should match the results snapshot', () => {
    expect(results).toMatchSnapshot();
  });

  it('should reject transactions that exceed the daily load count limit', () => {
    expect(results[3].accepted).toBe(false);
  });

  it('should reject transactions with a load_amount that exceeds the daily limit', () => {
    expect(results[4].accepted).toBe(false);
  });

  it('should reject duplicate transactions', () => {
    expect(results[7].accepted).toBe(false);
  });

  it("should reject transactions that would cause the customer's daily load amount to exceed the limit", () => {
    expect(results[8].accepted).toBe(false);
  });

  it("should reject transactions that would cause the customer's weekly load amount to exceed the limit", () => {
    expect(results[13].accepted).toBe(false);
  });

  it('should accept the other transactions', () => {
    results.forEach((result, i) => {
      if ([3, 4, 7, 8, 13].includes(i)) {
        return;
      }

      expect(result.accepted).toBe(true);
    });
  });
});
