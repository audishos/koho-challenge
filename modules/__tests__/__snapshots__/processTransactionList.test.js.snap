// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processTransactionList.test should match the results snapshot 1`] = `
Array [
  Object {
    "accepted": true,
    "customer_id": "1",
    "id": "1",
  },
  Object {
    "accepted": true,
    "customer_id": "1",
    "id": "2",
  },
  Object {
    "accepted": true,
    "customer_id": "1",
    "id": "3",
  },
  Object {
    "accepted": false,
    "customer_id": "1",
    "id": "4",
  },
  Object {
    "accepted": false,
    "customer_id": "8",
    "id": "420",
  },
  Object {
    "accepted": true,
    "customer_id": "40",
    "id": "666",
  },
  Object {
    "accepted": true,
    "customer_id": "40",
    "id": "675",
  },
  Object {
    "accepted": false,
    "customer_id": "40",
    "id": "675",
  },
  Object {
    "accepted": false,
    "customer_id": "40",
    "id": "686",
  },
  Object {
    "accepted": true,
    "customer_id": "123",
    "id": "728",
  },
  Object {
    "accepted": true,
    "customer_id": "123",
    "id": "734",
  },
  Object {
    "accepted": true,
    "customer_id": "123",
    "id": "756",
  },
  Object {
    "accepted": true,
    "customer_id": "123",
    "id": "757",
  },
  Object {
    "accepted": false,
    "customer_id": "123",
    "id": "758",
  },
  Object {
    "accepted": true,
    "customer_id": "5",
    "id": "1025",
  },
]
`;
